### PR TITLE
add data transformer structure

### PR DESCRIPTION
**Pull request by stegben**
pick from #26 .  Add data transformer

### CODE REVIEW

File: uttut/transformers/base.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    @abc.abstractmethod
+    def humanize(self, datum):
+        pass
+
+    @abc.abstractmethod
+    def machanize(self, raw_dict):
+        pass
+
+    @abc.abstractmethod
+    def serialize(self):
```
Comment (SoluMilken) [inline uttut/transformers/base.py]: save?

File: uttut/transformers/base.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def deserialize(cls, serialized):
```
Comment (SoluMilken) [inline uttut/transformers/base.py]: load ?

File: uttut/transformers/ordinal_label.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        else:
+            raise ValueError('Intent mapping is not valid.')
+
+        if OrdinalLabel.is_valid_mapping(entity2index):
+            self._entity2index = entity2index
+        else:
+            raise ValueError('Entity mapping is not valid.')
+
+    @staticmethod
+    def is_valid_mapping(str2idx: Mapping[str, int]) -> bool:
```
Comment (SoluMilken) [inline uttut/transformers/ordinal_label.py]: 要 private嗎？

File: uttut/transformers/ordinal_label.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            if idx == current_idx:
+                return intent
+        else:
+            raise KeyError("Intent index {} not found".format(idx))
+
+    def humanize(self, datum):
+        raw_dict = {}
+        raw_dict['utterance'] = datum.utterance
+        raw_dict['intent'] = {
+            'names': [self.index2intent(intent.idx) for intent in datum.intents],
```
Comment (SoluMilken) [inline uttut/transformers/ordinal_label.py]: intent.idx -> intent.label

File: uttut/transformers/ordinal_label.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        if datum.has_entities():
+            entities = []
+
+            for entity in datum.entities:
```
Comment (SoluMilken) [inline uttut/transformers/ordinal_label.py]: 有留entity 的 to_dict嗎？

**Comment by SoluMilken** [comment]
有點好奇 未來有可能有更多的transformer嗎？

File: uttut/transformers/ordinal_label.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    @classmethod
+    def deserialize(cls, serialized):
+        loaded = json.loads(serialized)
+        intent2index = loaded['intent2index']
+        entity2index = loaded['entity2index']
+        return cls(intent2index, entity2index)
+
+    @classmethod
+    def from_raw_dictionary(cls, raw_dict):
+        intents = [i for d in raw_dict['data'] for i in d['intent']['names']]
```
Comment (SoluMilken) [inline uttut/transformers/ordinal_label.py]: 為毛不set([i for d in raw_dict['data'] for i in d['intent']['names']])?

Comment (SoluMilken) [reply uttut/transformers/ordinal_label.py]:
```python
from collections import OrderedDict
A = OrderedDict.fromkeys([1, 2, 3, 1, 1, 2, 2, 3])
for a in A: print(a)
>
1
2
3
```

Comment (stegben) [reply uttut/transformers/ordinal_label.py]: 但這仍然需要過一層轉換。之後再想有沒有更簡潔的作法吧

**Comment by stegben** [comment]
Done, inject ignored entity from `__init__`

**Comment by stegben** [comment]
Done.

Creation:
While training
```python
from uttut.transformers import OrdinalLabel

raw_data = {
    "utterance": "我想喝珍奶半糖",
    "intent": {
        "names": ["ORDER"],
    },
    "entities": [
        {
            "name": "ITEM",
            "start": 3,
            "end": 5,
            "replacements": ["拿鐵", "多多綠", "紅茶"],
        },
        {
            "name": "SUGAR",
            "start": 5,
            "end": 7,
            "replacements": ["無糖", "全糖"],
        },
    ],
}
tx = OrdinalLabel.from_raw_dictionary(raw_data)

serialized_tx = tx.serialize()
with open('transformer.json', 'w') as fw:
    fw.write(serialized_tx)
```

At prediction time:
```python
from uttut.transformers import OrdinalLabel

with open('transformer.json', 'r') as f:
    serialized_tx = f.read()
tx = OrdinalLabel.load(serialized_tx)
```

**Comment by stegben** [comment]
To get `NOT_ENTITY` index, simply call property:
```python
tx.not_entity_index
```

File: uttut/transformers/ordinal_label.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            entity2index: Mapping[str, int],
+            not_entity_index: int = 0,
+        ):
+        if OrdinalLabel.is_valid_mapping(intent2index):
+            self._intent2index = intent2index
+        else:
+            raise ValueError('Intent mapping is not valid.')
+
+        self._not_entity_index = not_entity_index
+        if OrdinalLabel.is_valid_mapping({**entity2index, "__NOT_ENTITY__": not_entity_index}):
```
Comment (GBLin5566) [inline uttut/transformers/ordinal_label.py]:
```
__NOT_ENTITY__
```
這東西丟出去共用？

Comment (stegben) [reply uttut/transformers/ordinal_label.py]: ee8eaca

File: uttut/transformers/ordinal_label.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+        entity2index = {}
+        idx = 1
+        for entity in entities:
+            if entity not in entity2index:
+                entity2index[entity] = idx
+                idx += 1
+            continue
+
+        return cls(intent2index, entity2index, not_entity_index=0)
```
[the remark this answers is not in the document]
Comment (stegben) [reply uttut/transformers/ordinal_label.py]: 201f945

**Pull request merged**